### PR TITLE
Afform - Quick add links for Autocomplete fields

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
APIv3-based *EntityRef* fields have quick-add buttons which link to profile forms (e.g. "New Individual"). This implements a similar feature for the newer APIv4-based *Autocomplete* fields, using Afforms instead of Profiles. 

See https://lab.civicrm.org/dev/core/-/issues/4484

Usage
--------
1. Create a new form in FormBuilder.
2. Add an Autocomplete field for Contacts (so far this PR only includes "New Individual" and "New Organization" quick-add forms for testing).
3. Configure the field and choose "New Individual" or "New Organization" (or both) from the "Quick Add" dropdown.
image
4. Visit the form and you'll be able to add a new individual on-the-fly.

Technical Details
-------
For lack of anything fancier, the "quick-add" forms are identified by their route: any form with a route starting with "civicrm/quick-add" is automatically available in the dropdown pictured in step 3 (if the entity type matches).

This makes it quite configurable/flexible. You can make as many quick-add forms as you wish, to suit different types of Autocomplete fields.

### CODE REVIEW

**Comment by highfalutin** [comment]
I still want to advocate for allowing admins to override the default quick-add form. An example use case:

- I have an Activity submission form called "Prize Awarded"
- I have two entityref fields on the form: "Nomination" and "Reference Check". Both are to Activities -- but two different types of activities, each with their own custom fields.
- I'd like to be able to quick-add a _nomination_ and/or _reference check_ while I'm filling out the "Prize Awarded" form. The two quick-add forms would be different from each other.
    - each one sets a different activity type as a "value" (not exposed to the user)
    - each one includes just the custom fields  appropriate for its type

I think making this possible wouldn't be too hard. As you code the UI for configuring quick-add on entityref fields, you'd just need to provide options for "default" (which would be `civicrm/quick-add/Activity`), and also a drop-down menu allowing the admin to select any existing submission form route.

When it comes to rendering the afform, if non-default quick-add forms have been selected, they'd need to be prepended to `CRM.config.quickAdd`.

File: ext/afform/core/ang/af/afForm.component.js [hunk 1]
```diff
@@ -197,10 +199,9 @@
           }
         })
         .catch(function(error) {
-          status.resolve();
-          status = CRM.status(error.error_message, 'error');
+          status.reject();
           $element.unblock();
-          CRM.alert(error.error_message, ts('Form Error'));
+          CRM.alert(error.error_message || '', ts('Form Error'));
```
Comment (highfalutin) [inline ext/afform/core/ang/af/afForm.component.js:204]: changes in this `catch` block seem like they are pretty independent of the larger PR and could be merged quickly as a separate PR

File: ext/afform/core/ang/af/afForm.component.js [hunk 2]
```diff
@@ -112,7 +113,8 @@
 
         $element.trigger('crmFormSuccess', {
           afform: metaData,
-          data: data
+          data: data,
+          submissionResponse: submissionResponse,
```
[the remark this answers is not in the document]
Comment (colemanw) [reply ext/afform/core/ang/af/afForm.component.js:117]: Ok @highfalutin I've broken off as https://github.com/civicrm/civicrm-core/pull/27899

**Comment by colemanw** [comment]
@highfalutin I had to go chase a couple rabbits down their holes before coming back to this. But now that #27553 and #27659 and #27713 are merged I've rebased & completed work on this PR and it's working pretty well.

> I still want to advocate for allowing admins to override the default quick-add form.

Your wish is granted. 🧞 I'll update the description with a step-by-step. Let me know when you get a chance to review.

**Comment by colemanw** [comment]
Retest this please

**Comment by briennekordis** [comment]
When I click on the *Quick Add* drop down menu, I get 'None found', instead of 'New Individual'
Selection_051

**Comment by colemanw** [comment]
retest this please

**Comment by colemanw** [comment]
retest this please

**Comment by colemanw** [comment]
Thanks for testing @briennekordis I just rebuilt the demo site for this PR and now it's present:

image

**Comment by colemanw** [comment]
Wow Jenkins has really been having a bad week.
retest this please

**Comment by highfalutin** [comment]
Testing this locally, I am able to create an afform (A1), add an Existing Contact field, and select "New Individual" as the quick-add form for that field.

However, when I open the rendered afform A1 and click the "New Individual" button, the modal form that opens is another copy of the same form A1 -- not a New Individual form.

**Comment by colemanw** [comment]
Hmm, ok let's get a new online demo going. @civicrm-builder retest this please

File: js/Common.js [hunk 3]
```diff
@@ -628,6 +653,34 @@ if (!CRM.vars) CRM.vars = {};
               $el.select2('data', item, true);
             }
             return false;
+          })
+          // Pop-up Afform when clicking quick-add links
+          .on('click.crmEntity', '.crm-entityref-quick-add a', () => {
+            let url = $(this).attr('href');
```
Comment (highfalutin) [inline js/Common.js:659]:
```suggestion
            let url = $('#select2-drop .crm-entityref-quick-add a').attr('href');
```

Comment (highfalutin) [reply js/Common.js:659]: For some reason, in my testing, `this` was referring to the wrong element. Changing this line fixes the quick-add behavior.

**Review by highfalutin — CHANGES_REQUESTED** [review]

**Comment by highfalutin** [comment]
<details><summary>afforms for testing</summary>
<pre>
[
  {
    "requires": [],
    "title": "Event Registration with Quick Add",
    "description": "",
    "is_public": false,
    "permission": [
      "access CiviCRM"
    ],
    "type": "form",
    "entity_type": null,
    "join_entity": null,
    "placement": [],
    "summary_contact_type": null,
    "summary_weight": null,
    "icon": "fa-list-alt",
    "server_route": "civicrm/reg",
    "permission_operator": "AND",
    "redirect": null,
    "submit_enabled": true,
    "submit_limit": null,
    "create_submission": true,
    "navigation": {
      "parent": null,
      "label": "Event Registration with Quick Add",
      "weight": 0
    },
    "name": "afformEventRegistrationWithQuickAdd",
    "layout": [
      {
        "#tag": "af-form",
        "ctrl": "afform",
        "#children": [
          {
            "#text": "\n  "
          },
          {
            "#tag": "af-entity",
            "data": {
              "status_id": "1",
              "event_id": "1",
              "role_id": [
                "1"
              ]
            },
            "actions": {
              "create": true,
              "update": false
            },
            "type": "Participant",
            "name": "Participant1",
            "label": "Participant 1",
            "security": "RBAC"
          },
          {
            "#text": "\n  "
          },
          {
            "#tag": "fieldset",
            "af-fieldset": "Participant1",
            "class": "af-container",
            "af-title": "Participant 1",
            "#children": [
              {
                "#text": "\n    "
              },
              {
                "#tag": "af-field",
                "name": "contact_id",
                "defn": {
                  "input_attrs": {
                    "quickAdd": "civicrm/quick-add/Individual"
                  }
                }
              },
              {
                "#text": "\n    "
              },
              {
                "#tag": "button",
                "class": "af-button btn btn-primary",
                "crm-icon": "fa-check",
                "ng-click": "afform.submit()",
                "#children": [
                  {
                    "#text": "Submit"
                  }
                ]
              },
              {
                "#text": "\n  "
              }
            ]
          },
          {
            "#text": "\n"
          }
        ]
      },
      {
        "#text": "\n"
      }
    ]
  },
  {
    "requires": [],
    "title": "New Individual Phone Emal",
    "description": "",
    "is_public": false,
    "permission": [
      "access CiviCRM"
    ],
    "type": "form",
    "placement": [],
    "icon": "fa-list-alt",
    "server_route": "civicrm/add-individual-phone-email",
    "permission_operator": "AND",
    "submit_enabled": true,
    "create_submission": true,
    "name": "afformNewIndividualPhoneEmal",
    "layout": [
      {
        "#tag": "af-form",
        "ctrl": "afform",
        "#children": [
          {
            "#text": "\n  "
          },
          {
            "#tag": "af-entity",
            "data": {
              "contact_type": "Individual",
              "source": "New Individual Phone Emal"
            },
            "type": "Contact",
            "name": "Individual1",
            "label": "Individual 1",
            "actions": {
              "create": true,
              "update": true
            },
            "security": "RBAC"
          },
          {
            "#text": "\n  "
          },
          {
            "#tag": "fieldset",
            "af-fieldset": "Individual1",
            "class": "af-container",
            "af-title": "Individual 1",
            "#children": [
              {
                "#text": "\n    "
              },
              {
                "#tag": "af-field",
                "name": "first_name"
              },
              {
                "#text": "\n    "
              },
              {
                "#tag": "af-field",
                "name": "last_name"
              },
              {
                "#text": "\n    "
              },
              {
                "#tag": "div",
                "af-join": "Email",
                "af-repeat": "Add",
                "af-copy": "Copy",
                "min": "1",
                "#children": [
                  {
                    "#text": "\n      "
                  },
                  {
                    "#tag": "afblock-contact-email"
                  },
                  {
                    "#text": "\n    "
                  }
                ]
              },
              {
                "#text": "\n    "
              },
              {
                "#tag": "div",
                "af-join": "Phone",
                "af-repeat": "Add",
                "af-copy": "Copy",
                "min": "1",
                "#children": [
                  {
                    "#text": "\n      "
                  },
                  {
                    "#tag": "afblock-contact-phone"
                  },
                  {
                    "#text": "\n    "
                  }
                ]
              },
              {
                "#text": "\n  "
              }
            ]
          },
          {
            "#text": "\n  "
          },
          {
            "#tag": "button",
            "class": "af-button btn btn-primary",
            "crm-icon": "fa-check",
            "ng-click": "afform.submit()",
            "#children": [
              {
                "#text": "Submit"
              }
            ]
          },
          {
            "#text": "\n"
          }
        ]
      },
      {
        "#text": "\n"
      }
    ],
    "entity_type": null,
    "join_entity": null,
    "summary_contact_type": null,
    "summary_weight": null,
    "redirect": null,
    "submit_limit": null,
    "navigation": null
  }
]
</pre>
</details>

**Comment by highfalutin** [comment]
<details><summary>afforms for testing</summary>
<pre>
[
  {
    "requires": [],
    "title": "Event Registration with Quick Add",
    "description": "",
    "is_public": false,
    "permission": [
      "access CiviCRM"
    ],
    "type": "form",
    "entity_type": null,
    "join_entity": null,
    "placement": [],
    "summary_contact_type": null,
    "summary_weight": null,
    "icon": "fa-list-alt",
    "server_route": "civicrm/reg",
    "permission_operator": "AND",
    "redirect": null,
    "submit_enabled": true,
    "submit_limit": null,
    "create_submission": true,
    "navigation": {
      "parent": null,
      "label": "Event Registration with Quick Add",
      "weight": 0
    },
    "name": "afformEventRegistrationWithQuickAdd",
    "layout": [
      {
        "#tag": "af-form",
        "ctrl": "afform",
        "#children": [
          {
            "#text": "\n  "
          },
          {
            "#tag": "af-entity",
            "data": {
              "status_id": "1",
              "event_id": "1",
              "role_id": [
                "1"
              ]
            },
            "actions": {
              "create": true,
              "update": false
            },
            "type": "Participant",
            "name": "Participant1",
            "label": "Participant 1",
            "security": "RBAC"
          },
          {
            "#text": "\n  "
          },
          {
            "#tag": "fieldset",
            "af-fieldset": "Participant1",
            "class": "af-container",
            "af-title": "Participant 1",
            "#children": [
              {
                "#text": "\n    "
              },
              {
                "#tag": "af-field",
                "name": "contact_id",
                "defn": {
                  "input_attrs": {
                    "quickAdd": "civicrm/quick-add/Individual/with-phone-email"
                  }
                }
              },
              {
                "#text": "\n    "
              },
              {
                "#tag": "button",
                "class": "af-button btn btn-primary",
                "crm-icon": "fa-check",
                "ng-click": "afform.submit()",
                "#children": [
                  {
                    "#text": "Submit"
                  }
                ]
              },
              {
                "#text": "\n  "
              }
            ]
          },
          {
            "#text": "\n"
          }
        ]
      },
      {
        "#text": "\n"
      }
    ]
  },
  {
    "requires": [],
    "title": "New Individual Phone Email",
    "description": "",
    "is_public": false,
    "permission": [
      "access CiviCRM"
    ],
    "type": "form",
    "entity_type": null,
    "join_entity": null,
    "placement": [],
    "summary_contact_type": null,
    "summary_weight": null,
    "icon": "fa-list-alt",
    "server_route": "civicrm/quick-add/Individual/with-phone-email",
    "permission_operator": "AND",
    "redirect": null,
    "submit_enabled": true,
    "submit_limit": null,
    "create_submission": true,
    "navigation": null,
    "name": "afformNewIndividualPhoneEmal",
    "layout": [
      {
        "#tag": "af-form",
        "ctrl": "afform",
        "#children": [
          {
            "#text": "\n  "
          },
          {
            "#tag": "af-entity",
            "data": {
              "contact_type": "Individual",
              "source": "New Individual Phone Email"
            },
            "type": "Contact",
            "name": "Individual1",
            "label": "Individual 1",
            "actions": {
              "create": true,
              "update": true
            },
            "security": "RBAC"
          },
          {
            "#text": "\n  "
          },
          {
            "#tag": "fieldset",
            "af-fieldset": "Individual1",
            "class": "af-container",
            "af-title": "Individual 1",
            "#children": [
              {
                "#text": "\n    "
              },
              {
                "#tag": "af-field",
                "name": "first_name"
              },
              {
                "#text": "\n    "
              },
              {
                "#tag": "af-field",
                "name": "last_name"
              },
              {
                "#text": "\n    "
              },
              {
                "#tag": "div",
                "af-join": "Email",
                "af-repeat": "Add",
                "af-copy": "Copy",
                "min": "1",
                "#children": [
                  {
                    "#text": "\n      "
                  },
                  {
                    "#tag": "afblock-contact-email"
                  },
                  {
                    "#text": "\n    "
                  }
                ]
              },
              {
                "#text": "\n    "
              },
              {
                "#tag": "div",
                "af-join": "Phone",
                "af-repeat": "Add",
                "af-copy": "Copy",
                "min": "1",
                "#children": [
                  {
                    "#text": "\n      "
                  },
                  {
                    "#tag": "afblock-contact-phone"
                  },
                  {
                    "#text": "\n    "
                  }
                ]
              },
              {
                "#text": "\n  "
              }
            ]
          },
          {
            "#text": "\n  "
          },
          {
            "#tag": "button",
            "class": "af-button btn btn-primary",
            "crm-icon": "fa-check",
            "ng-click": "afform.submit()",
            "#children": [
              {
                "#text": "Submit"
              }
            ]
          },
          {
            "#text": "\n"
          }
        ]
      },
      {
        "#text": "\n"
      }
    ]
  }
]
</pre>
</details>

File: js/Common.js (hunk 3 above)
Comment (highfalutin) [reply js/Common.js:659]: ... except when multiple quick-add buttons are configured, it only respects the first one. Hmmmm.

Comment (colemanw) [reply js/Common.js:659]: @highfalutin Aha! I found "some reason" - it's because arrow functions [behave differently](https://stackoverflow.com/a/33308151/2918050) than regular functions with `this` binding. 
Fixed now, thanks for the hint.

Comment (highfalutin) [reply js/Common.js:659]: OK, this works for multiple quick-add buttons:

```suggestion
            let url = $(event.target).attr('href');
```

File: js/Common.js [hunk 4]
```diff
@@ -538,7 +538,23 @@ if (!CRM.vars) CRM.vars = {};
...
   }
 
-  function getStaticOptionMarkup(staticItems) {
+  function renderQuickAddMarkup(quickAddLinks) {
+    if (!quickAddLinks || !quickAddLinks.length) {
+      return '';
+    }
+    let markup = '<div class="crm-entityref-links crm-entityref-quick-add">';
+    CRM.config.quickAdd.forEach((link) => {
+      if (quickAddLinks.includes(link.path)) {
```
Comment (highfalutin) [inline js/Common.js:547]:
```suggestion
      if (_.includes(quickAddLinks, link.path)) {
```

File: js/Common.js [hunk 5]
```diff
@@ -559,9 +575,11 @@ if (!CRM.vars) CRM.vars = {};
     }
     select2Options = select2Options || {};
     return $(this).each(function() {
-      var $el = $(this).off('.crmEntity'),
-        staticItems = getStaticOptions(select2Options.static),
-        multiple = !!select2Options.multiple;
+      const $el = $(this).off('.crmEntity');
+      let staticItems = getStaticOptions(select2Options.static),
+        quickAddLinks = select2Options.quickAdd,
```
Comment (highfalutin) [inline js/Common.js:580]:
```suggestion
        quickAddLinks = select2Options.quickAdd ? select2Options.quickAdd.split(',') : [],
```

**Review by highfalutin — CHANGES_REQUESTED** [review]
These changes are also needed; otherwise buttons for all afforms under "civicrm/quick-add/X" are shown.

File: js/Common.js (hunk 4 above)
Comment (colemanw) [reply js/Common.js:547]: I've been trying to use more vanilla and less lo-dash recently, as that seems to be the direction the whole internet is headed. Did you find a bug here or is it just a style thing?

Comment (highfalutin) [reply js/Common.js:547]: It's a bug.

Comment (colemanw) [reply js/Common.js:547]: Ah, I think I found the cause - quickAddLinks is a string but was supposed to be an array. One sec...

Comment (highfalutin) [reply js/Common.js:547]: That's why I made the Line 580 suggestion.

**Comment by colemanw** [comment]
@highfalutin thanks for patiently testing this. I've fixed the array formatting, the `this` binding, also added a "New Organization" form and fixed both forms to make "Email" not required. Re-tested locally and pushed.
We're getting close!

**Comment by highfalutin** [comment]
Thanks @colemanw! The last bug we were talking about still needs to be fixed. At [js/Common.js line 580](https://github.com/civicrm/civicrm-core/pull/27754/files#r1378904880), `quickAddLinks` is still a string, not an array.

The reason I propose `_.includes()` in line 547 is that (Array.prototype.includes)[https://caniuse.com/array-includes] is not yet supported in most browsers.

**Comment by colemanw** [comment]
> @highfalutin `quickAddLinks` is still a string, not an array.

I think that might just be your existing test forms. If you regenerate those with the updated patch I've fixed the Afform GUI to always set that value as an array.

> [Array.prototype.includes](https://caniuse.com/array-includes) is not yet supported in most browsers.

I'm not quite sure what you mean... that link shows green for every browser except IE. I think we've pretty much given up on maintaining IE support.

**Comment by highfalutin** [comment]
Ah, ok, I re-tested and am no longer seeing the bug. My bad re: `includes`. I was confusing it with another thing that I was looking up at the same time.

**Comment by highfalutin** [comment]
The last test I did was whether quick-adds could be used within quick-adds. And [they can](http://core-27754-26y6d.test-1.civicrm.org:8001/civicrm/reg)! 

IMO this can be merged.

**Comment by colemanw** [comment]
Excellent
